guppylang/error.py

```diff
@@ -169,27 +180,40 @@ def format_source_location(
 def pretty_errors(f: FuncT) -> FuncT:
     """Decorator to print custom error banners when a `GuppyError` occurs."""
 
+    def hook(
+        excty: type[BaseException], err: BaseException, traceback: TracebackType | None
+    ) -> None:
+        """Custom `excepthook` that intercepts `GuppyExceptions` for pretty printing."""
+        # Fall back to default hook if it's not a GuppyException or we're missing an
+        # error location
+        if not isinstance(err, GuppyError) or err.location is None:
+            sys.__excepthook__(excty, err, traceback)
+            return
+
+        loc = err.location
+        file, line_offset = get_file(loc), get_line_offset(loc)
+        assert file is not None
+        assert line_offset is not None
+        line = line_offset + loc.lineno - 1
+        sys.stderr.write(
+            f"Guppy compilation failed. Error in file {file}:{line}\n\n"
+            f"{format_source_location(loc)}\n"
+            f"{err.__class__.__name__}: {err.get_msg()}\n",
+        )
+
     @functools.wraps(f)
     def pretty_errors_wrapped(*args: Any, **kwargs: Any) -> Any:
-        try:
-            return f(*args, **kwargs)
-        except GuppyError as err:
-            # Reraise if we're missing a location
-            if not err.location:
+        with exception_hook(hook):
+            try:
+                return f(*args, **kwargs)
+            except GuppyError as err:
+                # For normal usage, this `try` block is not necessary since the
+                # excepthook is automatically invoked when the exception (which is being
+                # reraised below) is not handled. However, when running tests, we have
+                # to manually invoke the hook to print the error message, since the
+                # tests always have to capture exceptions.
+                if "pytest" in sys.modules:
+                    hook(type(err), err, err.__traceback__)
```

Comment:
https://docs.pytest.org/en/latest/example/simple.html#pytest-current-test-environment-variable

this seems like a more stable test?